src/node/mod.rs

```diff
@@ -0,0 +1,226 @@
+    ///
+    /// _Overall type: A → 2 where `child`: A → 2_
+    ///
+    /// _Type inference will fail if children are not of the correct type._
+    #[allow(clippy::should_implement_trait)]
+    fn not(context: &mut Context<J>, child: &Self) -> Result<Self, types::Error> {
+        let unit = Self::unit(context);
+        let pair_child_unit = Self::pair(context, child, &unit)?;
+        let bit_true = Self::bit_true(context);
+        let bit_false = Self::bit_false(context);
+        let case_true_false = Self::case(context, &bit_true, &bit_false)?;
+
+        Self::comp(context, &pair_child_unit, &case_true_false)
+    }
+
+    /// Create a DAG that computes Boolean _AND_ of the `left` and `right` child.
+    ///
+    /// _Overall type: A → 2 where `left`: A → 2 and `right`: A → 2_
+    ///
+    /// _Type inference will fail if children are not of the correct type._
+    fn and(context: &mut Context<J>, left: &Self, right: &Self) -> Result<Self, types::Error> {
+        let iden = Self::iden(context);
+        let pair_left_iden = Self::pair(context, left, &iden)?;
+        let bit_false = Self::bit_false(context);
+        let drop_right = Self::drop_(context, right);
+        let case_false_right = Self::case(context, &bit_false, &drop_right)?;
+
+        Self::comp(context, &pair_left_iden, &case_false_right)
+    }
+
+    /// Create a DAG that computes Boolean _OR_ of the `left` and `right`.
+    ///
+    /// _Overall type: A → 2 where `left`: A → 2 and `right`: A → 2_
+    ///
+    /// _Type inference will fail if children are not of the correct type._
+    fn or(context: &mut Context<J>, left: &Self, right: &Self) -> Result<Self, types::Error> {
+        let iden = Self::iden(context);
+        let pair_left_iden = Self::pair(context, left, &iden)?;
+        let drop_right = Self::drop_(context, right);
+        let bit_true = Self::bit_true(context);
+        let case_right_true = Self::case(context, &drop_right, &bit_true)?;
+
+        Self::comp(context, &pair_left_iden, &case_right_true)
+    }
+}
+
+/// A node in a Simplicity expression.
+///
+/// There are several kinds of nodes...
+///
+/// For equality and hashing purposes, nodes are characterized entirely by their
+/// CMR and cached data. (For `Contruct` nodes this is, ...)
+#[derive(Debug)]
+pub struct Node<N: NodeData<J>, J: Jet> {
+    inner: Inner<Arc<Node<N, J>>, J, N::Witness>,
+    cmr: Cmr,
+    data: N::CachedData,
+}
+
+impl<N: NodeData<J>, J: Jet> PartialEq for Node<N, J>
+where
+    N::CachedData: PartialEq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.cmr == other.cmr && self.data == other.data
+    }
+}
+impl<N: NodeData<J>, J: Jet> Eq for Node<N, J> where N::CachedData: Eq {}
+
+impl<N: NodeData<J>, J: Jet> hash::Hash for Node<N, J>
+where
+    N::CachedData: hash::Hash,
+{
+    fn hash<H: hash::Hasher>(&self, h: &mut H) {
+        self.cmr.hash(h);
+        self.data.hash(h);
+    }
+}
```

Comment:
37a6c18ec24f06e9f6a2255b5cd529658f473abd: One downside of this generic approach is that things like unification arrows and node bounds influence equality. I feel like we should implement custom `Eq` implementations for `Final` and `{ConstructData,Commit,Redeem}Data`.

Reply:
I think that:

* `ConstructData` should not implement any equality or hashing traits
* `CommitData` should implement `PartialEq` which compares incomplete IMRs, but not `Eq` or `Hash`
* `RedeemData` should implement all three traits based on IMR

Does this sound reasonable?